infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java

```diff
@@ -141,280 +158,315 @@ private boolean isBootstrapped(CorfuMsg msg, ChannelHandlerContext ctx, IServerR
+        if(!isBootstrapped(req.getHeader(), ctx, r)) return;
 
-        final long msgEpoch = msg.getPayload();
+        final long payloadEpoch = req.getPayload().getLayoutRequest().getEpoch();
         final long serverEpoch = getServerEpoch();
 
-        if (msgEpoch <= serverEpoch) {
-            r.sendResponse(ctx, msg, new LayoutMsg(getCurrentLayout(), CorfuMsgType
-                    .LAYOUT_RESPONSE));
+        if(payloadEpoch <= serverEpoch) {
+            HeaderMsg responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            ResponseMsg response = getResponseMsg(responseHeader, getLayoutResponseMsg(getCurrentLayout()));
+            r.sendResponse(response, ctx);
         } else {
-            // else the client is somehow ahead of the server.
-            //TODO figure out a strategy to deal with this situation
-            r.sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_EPOCH, serverEpoch));
-            log.warn("handleMessageLayoutRequest: Message Epoch {} ahead of Server epoch {}",
-                    msgEpoch, serverEpoch);
+            //TODO: Client is ahead of the server. Is any other handling required?
+            log.warn("handleLayoutRequest[{}]: Payload epoch {} ahead of Server epoch {}",
+                    req.getHeader().getRequestId(), payloadEpoch, serverEpoch);
+
+            r.sendWrongEpochError(req.getHeader(), ctx, serverEpoch);
         }
     }
 
     /**
      * Sets the new layout if the server has not been bootstrapped with one already.
      *
-     * @param msg corfu message containing LAYOUT_BOOTSTRAP
+     * @param req corfu message containing BOOTSTRAP_LAYOUT_REQUEST
      * @param ctx netty ChannelHandlerContext
      * @param r   server router
      */
-    @ServerHandler(type = CorfuMsgType.LAYOUT_BOOTSTRAP)
-    public synchronized void handleMessageLayoutBootstrap(
-            @NonNull CorfuPayloadMsg<LayoutBootstrapRequest> msg,
-            ChannelHandlerContext ctx,
-            @NonNull IServerRouter r) {
-
-        if (getCurrentLayout() == null) {
-            Layout layout = msg.getPayload().getLayout();
-
-            log.info("handleMessageLayoutBootstrap: Bootstrap with new layout={}, {}",
-                    layout, msg);
-
-            if (layout.getClusterId() == null) {
-                log.warn("handleMessageLayoutBootstrap: The layout does {} not have a clusterId",
-                        layout);
-                r.sendResponse(ctx, msg, new CorfuMsg(CorfuMsgType.NACK));
-            }
-            else{
+    @RequestHandler(type = RequestPayloadMsg.PayloadCase.BOOTSTRAP_LAYOUT_REQUEST)
+    public synchronized void handleBootstrapLayoutRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        final HeaderMsg requestHeader = req.getHeader();
+        HeaderMsg responseHeader;
+        ResponseMsg response;
+
+        if(getCurrentLayout() == null) {
```

Comment:
Can you reverse the logic and get rid of the if-else here?

You can check `if(getCurrentLayout() != null)` you can send the error message and return right away. Otherwise continue to save the new bootstrap layout.